File: src/main/resources/public/ts/model/Resource.ts
export interface IResourceResponse {
    authors: string[];
    date: number;
    description: string;
    disciplines: string[];
    displayTitle: string;
    document_types: string[];
    editors: string[];
    hash: number;
    id: string;
    favorite?: boolean;
    image: string;
    levels: string[];
    link: string;
    plain_text: string[];
    source: string;
    title: string;
    action: object;
    structure_name: string;
    structure_uai: string;
    display_structure_name?: boolean;
    _id?: string;

    user?: string;
}

export class Resource {
    private _authors: string[];
    private _date: number;
    private _description: string;
    private _disciplines: string[];
    private _displayTitle: string;
    private _document_types: string[];
    private _editors: string[];
    private _hash: number;
    private _id: string;
    private _favorite?: boolean;
    private _image: string;
    private _levels: string[];
    private _link: string;
    private _plain_text: string[];
    private _source: string;
    private _title: string;
    private _action: object;
    private _structure_name: string;
    private _structure_uai: string;
    private _display_structure_name?: boolean;

    private _user?: string;

    constructor() {
        this._id = null;
        this._authors = null;
        this._date = null;
        this._description = null;
        this._disciplines = null;
        this._displayTitle = null;
        this._document_types = null;
        this._editors = null;
        this._hash = null;
        this._favorite = null;
        this._image = null;
        this._levels = null;
        this._link = null;
        this._plain_text = null;
        this._source = null;
        this._title = null;
        this._action = null;
        this._structure_name = null;
        this._structure_uai = null;
        this._display_structure_name = null;
        this._user = null;
    }

    build(data: IResourceResponse): Resource {
        this._id = data.id;
        this._authors = data.authors ? data.authors : [];
        this._date = data.date;
        this._description = data.description;
        this._disciplines = data.disciplines;
        this._displayTitle = data.displayTitle;
        this._document_types = data.document_types;
        this._editors = data.editors;
        this._hash = data.hash;
        this._favorite = data.favorite;
        this._image = data.image;
        this._levels = data.levels;
        this._link = data.link;
        this._plain_text = data.plain_text;
        this._source = data.source;
        this._title = data.title;
        this._action = data.action;
        this._structure_name = data.structure_name;
        this._structure_uai = data.structure_uai;
        this._display_structure_name = data.display_structure_name;
        this._user = data.user;

        return this;
    }

    get authors(): string[] {
        return this._authors;
    }

    set authors(value: string[]) {
        this._authors = value;
    }

    get date(): number {
        return this._date;
    }

    set date(value: number) {
        this._date = value;
    }

    get description(): string {
        return this._description;
    }

    set description(value: string) {
        this._description = value;
    }

    get disciplines(): string[] {
        return this._disciplines;
    }

    set disciplines(value: string[]) {
        this._disciplines = value;
    }

    get displayTitle(): string {
        return this._displayTitle;
    }

    set displayTitle(value: string) {
        this._displayTitle = value;
    }

    get document_types(): string[] {
        return this._document_types;
    }

    set document_types(value: string[]) {
        this._document_types = value;
    }

    get editors(): string[] {
        return this._editors;
    }

    set editors(value: string[]) {
        this._editors = value;
    }

    get hash(): number {
        return this._hash;
    }

    set hash(value: number) {
        this._hash = value;
    }

    get id(): string {
        return this._id;
    }

    set id(value: string) {
        this._id = value;
    }

    get favorite(): boolean {
        return this._favorite;
    }

    set favorite(value: boolean) {
        this._favorite = value;
    }

    get image(): string {
        return this._image;
    }

    set image(value: string) {
        this._image = value;
    }

    get levels(): string[] {
        return this._levels;
    }

    set levels(value: string[]) {
        this._levels = value;
    }

    get link(): string {
        return this._link;
    }

    set link(value: string) {
        this._link = value;
    }

    get plain_text(): string[] {
        return this._plain_text;
    }

    set plain_text(value: string[]) {
        this._plain_text = value;
    }

    get source(): string {
        return this._source;
    }

    set source(value: string) {
        this._source = value;
    }

    get title(): string {
        return this._title;
    }

    set title(value: string) {
        this._title = value;
    }

    get action(): object {
        return this._action;
    }

    set action(value: object) {
        this._action = value;
    }

    get structure_name(): string {
        return this._structure_name;
    }

    set structure_name(value: string) {
        this._structure_name = value;
    }

    get structure_uai(): string {
        return this._structure_uai;
    }

    set structure_uai(value: string) {
        this._structure_uai = value;
    }

    get display_structure_name(): boolean {
        return this._display_structure_name;
    }

    set display_structure_name(value: boolean) {
        this._display_structure_name = value;
    }

    get user(): string {
        return this._user;
    }

    set user(value: string) {
        this._user = value;
    }
}

File: backend/src/main/java/fr/openent/mediacentre/model/Resource.java
package fr.openent.mediacentre.model;

import fr.openent.mediacentre.core.constants.Field;
import fr.openent.mediacentre.helper.IModelHelper;
import fr.wseduc.mongodb.MongoDb;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.core.logging.Logger;
import io.vertx.core.logging.LoggerFactory;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import static fr.openent.mediacentre.core.constants.Field.*;

public abstract class Resource {
    protected static final Logger log = LoggerFactory.getLogger(Resource.class);
    private String _id;

    private String title;

    private List<String> plainText;

    private List<String> levels;

    private List<String> disciplines;

    private List<String> editors;

    private List<String> authors;

    private String image;

    private List<String> documentTypes;

    private String link;

    private String source;

    private boolean favorite;

    private JsonObject date;

    public Resource() {
    }

    public Resource(JsonObject resource) {
        this._id = resource.getString(_ID, null);
        this.title = resource.getString(TITLE, null);
        this.plainText = this.getStringListFromField(resource, PLAIN_TEXT);
        this.levels = this.getStringListFromField(resource, LEVELS);
        this.disciplines = this.getStringListFromField(resource, DISCIPLINES);
        this.editors = this.getStringListFromField(resource, EDITORS);
        this.authors = this.getStringListFromField(resource, AUTHORS);
        this.image = resource.getString(IMAGE, null);
        this.documentTypes = IModelHelper.toStringList(resource.getJsonArray(DOCUMENT_TYPES, new JsonArray()));
        this.link = resource.getString(LINK, null);
        this.source = resource.getString(SOURCE, null);
        this.favorite = resource.getBoolean(FAVORITE, false);
        this.date = this.formatDate(resource);
    }

    public String get_id() {
        return _id;
    }

    public Resource set_id(String _id) {
        this._id = _id;
        return this;
    }

    public String getTitle() {
        return title;
    }

    public Resource setTitle(String title) {
        this.title = title;
        return this;
    }

    public List<String> getPlainText() {
        return plainText;
    }

    public Resource setPlainText(List<String> plainText) {
        this.plainText = plainText;
        return this;
    }

    public List<String> getLevels() {
        return levels;
    }

    public Resource setLevels(List<String> levels) {
        this.levels = levels;
        return this;
    }

    public List<String> getDisciplines() {
        return disciplines;
    }

    public Resource setDisciplines(List<String> disciplines) {
        this.disciplines = disciplines;
        return this;
    }

    public List<String> getEditors() {
        return editors;
    }

    public Resource setEditors(List<String> editors) {
        this.editors = editors;
        return this;
    }

    public List<String> getAuthors() {
        return authors;
    }

    public Resource setAuthors(List<String> authors) {
        this.authors = authors;
        return this;
    }

    public String getImage() {
        return image;
    }

    public Resource setImage(String image) {
        this.image = image;
        return this;
    }

    public List<String> getDocumentTypes() {
        return documentTypes;
    }

    public Resource setDocumentTypes(List<String> documentTypes) {
        this.documentTypes = documentTypes;
        return this;
    }

    public String getLink() {
        return link;
    }

    public Resource setLink(String link) {
        this.link = link;
        return this;
    }

    public String getSource() {
        return source;
    }

    public Resource setSource(String source) {
        this.source = source;
        return this;
    }

    public boolean isFavorite() {
        return favorite;
    }

    public Resource setFavorite(boolean favorite) {
        this.favorite = favorite;
        return this;
    }

    public JsonObject getDate() {
        return date;
    }

    public Resource setDate(JsonObject date) {
        this.date = date;
        return this;
    }

    public JsonObject toJson() {
        return new JsonObject()
                .put(Field._ID, this.get_id())
                .put(Field.TITLE, this.getTitle())
                .put(Field.PLAIN_TEXT, new JsonArray(this.getPlainText()))
                .put(Field.LEVELS, new JsonArray(this.getLevels()))
                .put(Field.DISCIPLINES, new JsonArray(this.getDisciplines()))
                .put(Field.EDITORS, new JsonArray(this.getEditors()))
                .put(Field.AUTHORS, new JsonArray(this.getAuthors()))
                .put(Field.IMAGE, this.getImage())
                .put(Field.DOCUMENT_TYPES, new JsonArray(this.getDocumentTypes()))
                .put(Field.LINK, this.getLink())
                .put(Field.SOURCE, this.getSource())
                .put(Field.FAVORITE, this.isFavorite());
    }

    // Utils

    private List<String> getStringListFromField(JsonObject resource, String key) {
        try {
            // Case value is []
            return IModelHelper.toStringList(resource.getJsonArray(key, new JsonArray()));
        }
        catch (Exception e1) {
            try {
                // Case value is "[]"
                return IModelHelper.toStringList(new JsonArray(resource.getString(key, "")));
            }
            catch (Exception e2) {
                try {
                    // Case value is ""
                    return IModelHelper.toStringList(resource.getString(key, ""));
                }
                catch (Exception e3) {
                    log.error("[Mediacentre@Resource::getStringListFromField] error in converting data into list from key " + key);
                    return new ArrayList<>();
                }
            }
        }
    }

    private JsonObject formatDate(JsonObject resource) {
        Object value = resource.getValue(DATE, null);
        if (value == null) return MongoDb.now();

        if (value instanceof Long) return MongoDb.toMongoDate(new Date((Long) value));
        else if (value instanceof JsonObject) return (JsonObject) value;
        else return null;
    }
}